subprojects/GOPATH/src/github.com/microsoft/hivedscheduler/pkg/algorithm/hived_algorithm.go

```diff
@@ -88,68 +97,34 @@ func (h *HivedAlgorithm) Schedule(pod *core.Pod, suggestedNodes []string) intern
 
 	klog.Infof("[%v]: Scheduling pod...", internal.Key(pod))
 	s := internal.ExtractPodSchedulingSpec(pod)
+	// gpu number -> a set of pods -> a set of GPUs of each pod
+	var groupPhysicalPlacement map[int32][]CellList
+	var groupVirtualPlacement map[int32][]CellList
+	newGroup := true
+	podIndex := int32(0)
 
-	// Differences between group cells and pod cells:
-	// A group cell is allocated by the HiveD algorithm to an affinity group,
-	// and is the unit of HiveD's cell allocation.
-	// A pod cell is *only* for finding a GPU placement in a group cell for a specific pod.
-	// Pod placement operates on cells because this way we can use buddy alloc to avoid
-	// placing a pod cross multiple nodes.
-	// A pod cell is NOT for the cell allocation process!
-	if groupPhysical, groupVirtual := h.getOrRequestGroupCell(pod, s); groupPhysical != nil {
-		if podCell := h.requestPodCell(s, groupPhysical); podCell == nil {
-			return internal.PodScheduleResult{
-				PodPreemptInfo: &internal.PodPreemptInfo{},
-			}
-		} else {
-			nodes, gpuIndices := podCell.GetPhysicalPlacement()
-			if len(nodes) > 1 {
-				// TODO: request larger group cell (when scheduling the first pod),
-				//  or prove this is an impossibility
-				panic(fmt.Sprintf("pod placed across nodes: %v", podCell.GetName()))
-			}
-			selectedNode := nodes[0]
-			if !common.StringsContains(suggestedNodes, selectedNode) {
-				if !groupPhysical.HasPod() {
-					panic(fmt.Sprintf(
-						"[%v]: node %v picked by algorithm but not in K8S candidates", internal.Key(pod), selectedNode))
-				}
-			}
-			groupNodes, groupGpu := groupPhysical.GetPhysicalPlacement()
-			vcl, vci := int32(-1), int32(-1)
-			if groupVirtual != nil {
-				vcl = int32(groupVirtual.GetLevel())
-				vci = groupVirtual.GetIndex()
-			}
-			return internal.PodScheduleResult{
-				PodBindInfo: &api.PodBindInfo{
-					Node:         selectedNode,
-					GpuIsolation: gpuIndices[:s.GpuNumber],
-					CellChain:    string(groupPhysical.GetChain()),
-					PodCellLevel: int32(podCell.GetLevel()),
-					PodCellGpuPlacement: api.Range{
-						Start: gpuIndices[0],
-						End:   gpuIndices[len(gpuIndices)-1]},
-					GroupCellLevel: int32(groupPhysical.GetLevel()),
-					GroupCellNodes: groupNodes,
-					GroupCellGpuPlacement: api.Range{
-						Start: groupGpu[0],
-						End:   groupGpu[len(groupGpu)-1],
-					},
-					VirtualCellLevel: vcl,
-					VirtualCellIndex: vci,
-				},
-			}
+	if group := h.allocatedAffinityGroups[s.AffinityGroup.Name]; group == nil {
+		klog.Infof("Scheduling new affinity group %v", s.AffinityGroup.Name)
+		groupPhysicalPlacement, groupVirtualPlacement = h.scheduleNewAffinityGroup(pod, s)
+		if groupPhysicalPlacement != nil {
 		}
```

Comment:
Empty if?

Reply:
it's a mistake. I forgot to delete it :)

---
In reply to: [335884633](https://github.com/microsoft/pai/pull/3684#discussion_r335884633) [](ancestors = 335884633)